### PR TITLE
nixos/healthchecks: add settingsFile option

### PR DESCRIPTION
**Pull request by yomaq**
Single unified env file read by systemd

## Description of changes

Current settings option requires multiple files if you want to set multiple environment secrets, as well as being read by the service program, meaning that things like agenix won't work with the default service user.

Added a settingsFile option that allows a single secret file, and doesn't break any current behavior for existing settings options, and plays well with agenix. Env variables can be set in both settings and settingsFile.

Tested both the webservice, and the healtchecks-manage script created by the module.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/web-apps/healthchecks.nix [hunk 1]
```diff
@@ -109,9 +116,11 @@ in
           have support for a `_FILE` variant, run:
           - `nix-instantiate --eval --expr '(import <nixpkgs> {}).healthchecks.secrets'`
           - or `nix eval 'nixpkgs#healthchecks.secrets'` if the flake support has been enabled.
+
+        If the same variable is set in both `settings` and `settingsFile` the value from `settingsFile` has priority.
       '';
       type = types.submodule (settings: {
-        freeformType = types.attrsOf types.str;
+        freeformType = types.attrsOf (types.nullOr types.str);
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/healthchecks.nix:123]: This should also already allow {} which is a bit better than null.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Why are multiple files required? The environmentFile by systemd can hold multiple envs already.

File: nixos/modules/services/web-apps/healthchecks.nix (hunk 1 above)
Comment (amozeo) [reply nixos/modules/services/web-apps/healthchecks.nix:123]: This is a change that I proposed in https://github.com/NixOS/nixpkgs/pull/314173#discussion_r1613159749
see the comment before it for reason why

**Comment by amozeo** [comment]
It should be noted that pr author originally created https://github.com/NixOS/nixpkgs/pull/314173, but wanted to rename the branch as stated in his last comment there

**Comment by yomaq** [comment]
Previously the module would have you set an env variable that would point to a file for each secret you wanted to set, and the program would read the secret file rather than systemd.

Only one file is required. The `settingsFile` option I can handle all secret requirements just fine, as it is using the systemd environmentFile directly.
But as to not to make any breaking changes I left the previous env variable options as is.

File: nixos/modules/services/web-apps/healthchecks.nix (hunk 1 above)
Comment (yomaq) [reply nixos/modules/services/web-apps/healthchecks.nix]: Removed the change as suggested.
The script issue from [#314173 (comment)](https://github.com/NixOS/nixpkgs/pull/314173#discussion_r1613159749) is fixed by line 133 where `settings.SECRET_KEY_FILE` can still be set as null.

File: nixos/modules/services/web-apps/healthchecks.nix [hunk 2]
```diff
@@ -186,7 +196,10 @@ in
           WorkingDirectory = cfg.dataDir;
           User = cfg.user;
           Group = cfg.group;
-          EnvironmentFile = [ environmentFile ];
+          EnvironmentFile = [
+            environmentFile
+            cfg.settingsFile
```
Comment (eclairevoyant) [inline nixos/modules/services/web-apps/healthchecks.nix:201]: missing null check?

Comment (yomaq) [reply nixos/modules/services/web-apps/healthchecks.nix]: It was, thank you

File: nixos/modules/services/web-apps/healthchecks.nix [hunk 3]
```diff
@@ -186,7 +196,10 @@ in
           WorkingDirectory = cfg.dataDir;
           User = cfg.user;
           Group = cfg.group;
-          EnvironmentFile = [ environmentFile ];
+          EnvironmentFile = [
+            environmentFile
+            (lib.optional (cfg.settingsFile != null) cfg.settingsFile)
+          ];
```
Comment (amozeo) [inline nixos/modules/services/web-apps/healthchecks.nix:202]: `lib.optional` returns a list
```suggestion
          EnvironmentFile = [
            environmentFile
          ] ++ lib.optional (cfg.settingsFile != null) cfg.settingsFile;
```

Comment (yomaq) [reply nixos/modules/services/web-apps/healthchecks.nix]: Module was working as it was, but it does return I list. I've fixed this now.

**Review by SuperSandro2000 — APPROVED** [review]

**Pull request merged**